### ISSUE TITLE
Sync is over-writting previously saved Data in iCloud instead of fetching it.

### ISSUE THREAD

**Issue by omarojo** [ArtSabintsev/Zephyr#35]
Device1, saved a bunch of data in iCloud. 
Device2, just installed the app, saved some default data locally during initialization, and then called Sync, expecting to download whatever was in iCloud and override the default data saved a moment before.  

What happened: Device1 now has the data of Device2 and I lost all data tha Device1 had previously uploaded to iCloud. I understand you are syncing with the Latest data, in this case Device2 pushed the latest data. But then how can I fetch what is in iCloud before saving defaults locally (specially when no internet) and not deleting everyhing in the cloud when calling sync.  

is Sync() a synchronous process? can I know if it failed? 

Any suggestion is appreciated. 
Thank you :)

**Comment by omarojo** [comment]
I see you mention something about Registering UserDefaults first. I fail to understand how that works  behind the scenes. Should I save my defaults with that method, and then after calling Zephr.sync() it wont take those ase the **latest** version to be uploaded and instead it will download what is already in iCloud ? or what is does that register() method do in relation with Zephr.

**Comment by ArtSabintsev** [comment]
Hi,

Registering `UserDefaults` allows you to run your app with some default information while you wait for the sync to happen. From https://developer.apple.com/documentation/foundation/userdefaults/1417065-register

```
The contents of the registration domain are not written to disk; you need to call this method each time your application starts. You can place a plist file in the application's Resources directory and call register(defaults:) with the contents that you read in from that file.
```

Once you perform some task in the app that requires a change to the registered `UserDefaults`, you save the defaults using the normal method so they persist across shutdowns. 

The assumption here is that if you have defaults stored somewhere from a different device, that iCloud will sync them quickly, which it usually does. 

No, there is no error handling or any of that, since this is a KVO API. Either you get data back, or you don't. 

Also, `sync()` just stand for `synchronize`, not `synchronous`.

So in your app:

Device 1: Launch app and register some defaults on launch
Device 1: Perform some actions that then set default values to one or more defaults
Device 1: Those values get synced to iCloud
Device 2: Launch app and register some defaults on launch
Device 2: Zephyr monitors for changes on launch
Device 2: Zephyr brings in those changes and automatically sets/saved to disk the synchronized values from Device-1/iCloud 

I hope that helps.

**Comment by omarojo** [comment]
Got it. 

In my experiment I was modifying UserDefaults on Device2 as soon as the viewController loaded, causing the sync() (called in **didFinishLaunchingWithOptions**) to recognize it as the new Master and uploading that to iCloud as the newest. ( I removed that, and now it worked, it pulled from iCloud first as long as I didnt modify anything before the Sync happened) 

So.... I also tried Installing the app from scratch in Device2 with **NO INTERNET** and making modifications to UserDefaults. When internet came back, no automatic sync happened, untili the next time I made a modification to UserDefaults, then boom...  it synced replaced everything that was in iCloud with Device2 data. Im guessing that is the expected behaviour. Since it didnt have a chance to sync() before the first modification, making any further mods as the new Master Data. 

I guess this is not the best way to sync data between devices, serves better as a mean to save data in iCloud for 1 device only, it stays consistent.
